mono/sgen/sgen-marksweep.c

```diff
@@ -693,6 +694,56 @@ major_alloc_object (GCVTable vtable, size_t size, gboolean has_references)
 }
 
 /*
+ * This can only be called by sgen workers. While this is called we assume
+ * that no other thread is accessing the block free lists. The world should
+ * be stopped and the gc thread should be waiting for workers to finish.
```

Comment:
As of right now yes. The gc thread cannot call this function because it lacks the tls storage for the private block free list, but it's not hard to add support. I'd rather keep it this way though, having gc thread and workers run in separate stages, so we are keeping things simple.